gremlin-scala/src/main/scala/gremlin/scala/ScalaElement.scala

```diff
@@ -14,17 +16,18 @@ trait ScalaElement[ElementType <: Element] {
 
   def keys: Set[String] = element.keys.toSet
 
+  def setProperty(key: String, value: Any): Any
+
   def property[A](key: String): Property[A] = element.property[A](key)
 
-  def properties(wantedKeys: String*): Seq[Property[Any]] = {
-    val requiredKeys = if (wantedKeys.nonEmpty) wantedKeys else keys
-    requiredKeys map property[Any]
-  }.toSeq
+  def properties[A]: Stream[Property[A]]
+
+  def properties[A](wantedKeys: String*): Stream[Property[A]]
 
-  def propertyMap(wantedKeys: String*): Map[String, Any] = {
-    val requiredKeys = if (wantedKeys.nonEmpty) wantedKeys else keys
-    requiredKeys map { key ⇒ (key, getValue(key)) }
-  }.toMap
+  def valueMap[A]: Map[String, A] = valueMap[A](keys.toSeq: _*)
+
+  def valueMap[A](wantedKeys: String*): Map[String, A] =
```

Comment:
The goal here is to give the optional ability to set the return type if the user wants:
`valueMap("someInteger", "someString")` should return a `Map[String, Any]`
`valueMap[String]("someString", "someString2")` should return a `Map[String, String]`

This is currently returning `Map[String, Nothing]` instead of `Map[String, Any]` by default but I have a fix for that, that I will push this evening.